Add topk operator

mgehre-amd opened the pull request:
Adds a topk operator

mgehre-amd committed:
Add topk operator

josel-amd committed:
Add the Indice to the outputs

Keep consistency with the onnx.topk

cferry-AMD commented:
I'm wondering whether we should also infer the type of `indices` in `inferReturnTypeComponents`?

mgehre-amd commented on include/xten/Dialect/XTenNN/IR/XTenNNOps.td:
```diff
@@ -512,6 +512,26 @@ def XtenNN_SignOp: XTenNN_Op<"sign", [Pure, TosaExtension, ElementwiseUnary, Sam
+  }];
+  let arguments = (ins
+    AnyRankedTensor:$input,
+    I64Attr:$k,
```
To be consistent with ONNX, this would need to be a Value instead of an Attr. ONNX wants this to be a `tensor<1xi64>`, but an `i64` value would also be fine for me

josel-amd replied:
Currently working on that :)

josel-amd committed:
Use K as an operand instead of attribute

Set k as an operand

josel-amd committed:
Do not make constant k mandatory

josel-amd committed:
Test topk with an argument

cferry-AMD commented on lib/Dialect/XTenNN/IR/XTenNNOps.cpp:
```diff
@@ -551,24 +556,21 @@ LogicalResult TopK::inferReturnTypeComponents(
 
   SmallVector<int64_t> resultShape{inTy.getShape()};
-  resultShape[axis] = *k;
+  resultShape[axis] = k ? *k : ShapedType::kDynamic;
```
I think this is the way to go for a dynamic shape :)

josel-amd committed:
Add support for dynamic topk

josel-amd committed:
k must be <= dimSize

josel-amd committed:
Remove unused include

josel-amd committed:
Relax result type comparison

Check if the types are compatible insteal of equal. Sometimes the result
type can be made static and type inference cannot deduce it e.g. when
the k is dynamic

josel-amd committed:
Add a test case

mgehre-amd commented:
LGTM

josel-amd committed:
Fix failing test

cferry-AMD reviewed (approved):
Thanks!
